[#1] 그룹 채팅방 생성

ByulKim opened the pull request:
<br/>

## 작업 내용
새로운 그룹 채팅방을 생성하는 API 개발 작업입니다.
( 관련 이슈 : #1 )

<br/>

## 그룹 채팅방 생성 API
- URL
    - POST /group/room

- request body
    - `roomOwnerId` (현재 그룹 채팅방 생성을 요청하는 사용자 ID)
    - `roomName` (그룹 채팅방 이름)
    - `enterCode` (그룹 채팅방 입장 코드)

- response body
    - `id` (생성된 채팅방 ID)
    - `type` (생성된 채팅방의 타입 - G:그룹채팅, M:마니또채팅)
    - `groupRoomDetail`
        - `roomId` (생성된 채팅방 ID)
        - `roomName` (그룹 채팅방 이름)
        - `roomOwnerId` (현재 그룹 채팅방 생성을 요청하는 사용자 ID)
        - `enterCode` (그룹 채팅방 입장 코드)
        - `createdAt` (생성일자)
        - `updatedAt` (수정일자)
    - `createdAt` (생성일자)
    - `updatedAt` (수정일자)

<br/>

## 테스트

1. Repository: RoomRepositoryTest, GroupRoomDetailRepository
2. Service: GroupChatServiceTest
3. Controller: GroupChatControllerTest

- Repository 와 Service는 로컬환경에서 mysql 을 설치하여 연결 후에
   실제 데이터 추가되는 것까지 포함하여 테스트 코드를 작성하였습니다.

<br/>

## 고민사항/질문

1. response body 부분도 필요할 것 같은 데이터만 골라 DTO로 생성하는것이 좋을까요?
2. repository, service 클래스의 테스트 코드 작성시에 DB에 접근하고 저장하는 부분 없이 로직만을 검증하도록 해야할까요?

<br/>

ByulKim committed:
mysql 설정정보 추가

ByulKim committed:
rooms, group_room_details 테이블 entity, repository 추가

ByulKim committed:
그룹 채팅방 생성 요청 DTO 추가 및 유효성 검사 Global Exception 추가

ByulKim committed:
그룹 채팅방 생성 API 추가
- POST /group/room
- 그룹채팅방의 이름, 방장 ID, 입장코드 전달 필요

ByulKim committed:
그룹 채팅방 생성 API 테스트 코드 추가

jinspark-lab commented on web/src/main/java/com/example/web/enums/RoomType.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.web.enums;
+
+public enum RoomType {
+    G, M
```
줄임말을 enum 으로 쓰신다면 description 같은거 남겨주시면 좋습니다.

G("Group"),
M("Manito"),
;

private String description;

jinspark-lab commented on web/src/test/java/com/example/web/repository/GroupRoomDetailRepositoryTest.java:
```diff
@@ -0,0 +1,43 @@
+        groupRoomDetail.setRoomOwner(roomOwnerId);
+        groupRoomDetail.setRoomName(roomName);
+        groupRoomDetail.setEnterCode(enterCode);
+        GroupRoomDetail newGroupRoomDetail = groupRoomDetailRepository.save(groupRoomDetail);
```
Mocking 해서 테스트하시는게 좋습니다.

jinspark-lab commented on web/src/main/java/com/example/web/controller/GroupChatController.java:
```diff
@@ -0,0 +1,26 @@
+    private final GroupChatService groupChatService;
+
+    @PostMapping("/room")
+    public ResponseEntity<Room> createGroupRoom(@Valid @RequestBody CreateGroupRoomDto dto) {
```
DTO 로 바꿔주시는게 좋습니다.

jinspark-lab commented on web/src/main/java/com/example/web/service/GroupChatService.java:
```diff
@@ -0,0 +1,32 @@
+public class GroupChatService {
+
+    private final RoomRepository roomRepository;
+    private final GroupRoomDetailRepository groupRoomDetailRepository;
```
Service 에서 다른 Service 를 호출하는게 좋습니다.
(예시)  GroupRoomDetailRepository 는 GroupRoomDetailService 를 통해서만 접근되는게 좋습니다

jinspark-lab commented on web/src/main/java/com/example/web/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,28 @@
+            String fieldName = ((FieldError) error).getField();
+            String errorMessage = error.getDefaultMessage();
+            errors.put(fieldName, errorMessage);
+        });
```
Reflection 은 지양해주세요

ByulKim replied:
@jinspark-lab 
`exception.getBindingResult().getAllErrors()`
피드백 주신 이 부분 수정하면서 궁금한 점이 있어 질문 드립니다!

`exception.getBindingResult()` 를 통해 현재 발생한 오류를 가져오고,
이 오류들 중에서 `result.getAllErrors()` 를 사용해 필드/메시지 조합으로 Body를 생성하는 코드인데
여기서 exception 객체를 통해 바인딩된 에러를 조회하는것이 문제였던 걸까요?!
에러를 조회하는 것만으로도 리플렉션으로 문제가 된 것인지 궁금해서 여쭤봅니다!

ByulKim committed:
RoomType enum 클래스 description 추가

ByulKim committed:
그룹 채팅방 생성 API 수정
- 컨트롤러, 서비스 클래스명 수정
- 서비스/레포지토리 클래스 분리
- DTO, VO 추가

ByulKim committed:
그룹 채팅방 생성 API 관련 단위 테스트 추가
- RoomRepositoryTest, GroupRoomDetailRepositoryTest
- GroupRoomServiceTest, GroupRoomDetailServiceTest
- GroupRoomControllerTest

ByulKim committed:
유효성검사 GlobalExceptionHandler 수정, 테스트 코드 추가
- GroupRoomControllerTest

ByulKim committed:
그룹 채팅방 정보 저장 로직 수정

ByulKim committed:
그룹 채팅방 생성 API 유효성 검사, exception handler 수정
